drivers/adc/adc_npcx.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2020 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_adc
+
+#include <assert.h>
+#include <drivers/adc.h>
+#include <drivers/clock_control.h>
+#include <kernel.h>
+#include <soc.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_npcx, CONFIG_ADC_LOG_LEVEL);
+
+/* ADC speed/delay values during initialization */
+#define ADC_REGULAR_DLY_VAL     0x03
+#define ADC_REGULAR_ADCCNF2_VAL 0x8B07
+#define ADC_REGULAR_GENDLY_VAL  0x0100
+#define ADC_REGULAR_MEAST_VAL   0x0001
+
+/* ADC channel number */
+#define NPCX_ADC_CH_COUNT 10
+
+/* ADC targeted operating frequency (2MHz) */
+#define NPCX_ADC_CLK 2000000
+
+/* ADC internal reference voltage (Unit:mV) */
+#define NPCX_ADC_VREF_VOL 2816
+
+/* ADC conversion mode */
+#define NPCX_ADC_CHN_CONVERSION_MODE  0
+#define NPCX_ADC_SCAN_CONVERSION_MODE 1
+
+/* Device config */
+struct adc_npcx_config {
+	/* adc controller base address */
+	uintptr_t base;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const struct npcx_alt *alts_list;
+};
+
+/* Driver data */
+struct adc_npcx_data {
+	/* Input clock for ADC converter */
+	uint32_t input_clk;
+	/* mutex of ADC channels */
+	struct adc_context ctx;
+	/*
+	 * Bit-mask indicating the channels to be included in each sampling
+	 * of this sequence.
+	 */
+	uint16_t channels;
+	/* Device pointer to get HAL instance */
```

Comment:
I don't think it is a HAL. It is just the chip registers.

Reply:
Okay. I will modify it.

Reply:
Done.